Sources/main.cpp

```diff
@@ -101,6 +104,150 @@ void GetArchiveFileList(const char *path,
     }
 }
 
+class MalformedMappingFileError : public std::exception
+{
+public:
+    explicit MalformedMappingFileError(off_t lineNumber)
+        : lineNumber(lineNumber)
+    {
+        this->SetFileName(nullptr);
+    }
+
+    const char *what() const noexcept override
+    {
+        return this->message.c_str();
+    }
+
+    void SetFileName(const char *fileName)
+    {
+        if (!fileName || strcmp(fileName, "") == 0) {
+            fileName = "(unknown)";
+        }
+        std::ostringstream ss;
+        ss << "Malformed mapping file: " << fileName << ":" << this->lineNumber;
+        this->message = ss.str();
+    }
+
+private:
+    std::string message;
+    off_t lineNumber;
+};
+
+bool GetLine(std::istream &file, std::string &out, char delimiter)
+{
+    try {
+        std::getline(file, out, delimiter);
+    } catch (std::ios_base::failure &e) {
+        if (file.fail() || file.eof()) {
+            // Handled below.
+        } else {
+            throw;
+        }
+    }
+    if (file.bad()) {
+        return false;
+    }
+    if (file.eof()) {
+        if (file.fail() && out.empty()) {
+            // Don't treat starting at EOF as a failure.
+            file.clear(file.rdstate() & ~std::istream::failbit);
+            return false;
+        } else {
+            // We parsed something and reached EOF. Process the line.
+            return true;
+        }
+    }
+    return true;
+}
+
+void GetArchiveFileListFromMappingFile(
+    std::istream &mappingFile,
+    std::unordered_map<std::string, std::string> &fileNames)
+{
+    static const char kWhitespace[] = " \t";
+    // TODO(strager): Make this parser more accepting. This parser is way too
+    // strict.
+    bool didReadHeader = false;
+    off_t lineNumber = 1;
+    for (;;) {
+        std::string line;
+        if (!GetLine(mappingFile, line, '\n')) {
+            break;
+        }
+        if (mappingFile.fail()) {
+            // The line is too long.
+            throw MalformedMappingFileError(lineNumber);
+        }
+
+        // Trim leading and trailing whitespace and ignore blank lines.
+        {
+            auto first = line.find_first_not_of(kWhitespace);
+            if (first == std::string::npos) {
+                // Blank line.
+                continue;
+            }
+            auto last = line.find_last_not_of(kWhitespace);
+            assert(last != std::string::npos);
+            line.erase(last + 1, std::string::npos);
+            line.erase(0, first);
+        }
+        if (didReadHeader) {
+            // Parse the following:
+            //
+            //     "localPath" "archiveName"
```

Comment:
> Is there a max path size?

No.

> I'm wondering whether this could be a single `sscanf`.

A proper parser would be better. =S I was just lazy.

Do you have a problem with the code as written? I know it's not very good, but I've tried to make it clear. Any incremental suggestions? (I'll accept a working rewrite too. =])